BLE_GattClient/source/BLEProcess.h

```diff
@@ -139,91 +146,104 @@ class BLEProcess : private mbed::NonCopyable<BLEProcess> {
 
     /**
      * Start the gatt client process when a connection event is received.
+     * This is called by Gap to notify the application we connected
      */
-    void when_connection(const Gap::ConnectionCallbackParams_t *connection_event)
-    {
-        printf("Connected.\r\n");
+    virtual void onConnectionComplete(
+        const ble::ConnectionCompleteEvent &event
+    ) {
+        if (event.getStatus() == BLE_ERROR_NONE) {
+            printf("Connected.\r\n");
+        } else {
+            printf("Failed to connect\r\n");
+            _event_queue.call(this, &BLEProcess::start_advertising);
+        }
     }
 
     /**
      * Stop the gatt client process when the device is disconnected then restart
      * advertising.
+     * This is called by Gap to notify the application we disconnected
      */
-    void when_disconnection(const Gap::DisconnectionCallbackParams_t *event)
-    {
+    virtual void onDisconnectionComplete(
+        const ble::DisconnectionCompleteEvent &event
+    ) {
         printf("Disconnected.\r\n");
-        start_advertising();
-    }
-
-    /**
-     * Setup the advertising parameters.
-     */
-    void setup_advertising()
-    {
-        Gap &gap = _ble_interface.gap();
-        gap.setAdvertisingPayload(make_advertising_data());
-        gap.setAdvertisingParams(make_advertising_params());
+        _event_queue.call(this, &BLEProcess::start_advertising);
     }
 
     /**
      * Start the advertising process; it ends when a device connects.
      */
     void start_advertising()
     {
-        ble_error_t error = _ble_interface.gap().startAdvertising();
-        if (error) {
-            printf("Error %u during gap.startAdvertising.\r\n", error);
-        } else {
-            printf("Advertising started.\r\n");
-        }
-    }
+        ble_error_t error;
 
-    /**
-     * Schedule processing of events from the BLE middleware in the event queue.
-     */
-    void schedule_ble_events(BLE::OnEventsToProcessCallbackContext *event)
-    {
-        _event_queue.call(mbed::callback(&event->ble, &BLE::processEvents));
-    }
+        ble::AdvertisingParameters adv_params;
+        adv_params.setUseLegacyPDU(false);
+        adv_params.setType(ble::advertising_type_t::CONNECTABLE_UNDIRECTED);
+        adv_params.setPrimaryInterval(
+            ble::periodic_interval_t(500), ble::periodic_interval_t(500)
+        );
 
-    /**
-     * Build data advertised by the BLE interface.
-     */
-    static GapAdvertisingData make_advertising_data(void)
-    {
-        static const uint8_t device_name[] = "GattClient";
-        GapAdvertisingData advertising_data;
+        /* create the advertising set with its parameter if we haven't yet */
+        if (_adv_handle == ble::INVALID_ADVERTISING_HANDLE) {
```

Comment:
We can use the legacy advertising set to advertise.

Reply:
Can you be a bit more explicit on what you are suggesting here?
Thanks.

Reply:
You don't need to create an advertising set, you can use the legacy one that is always present: `LEGACY_ADVERTISING_HANDLE ` ([reference](https://github.com/ARMmbed/mbed-os/blob/9243abb6be0cb43ffc40df675cbdd443a796ff66/features/FEATURE_BLE/ble/gap/Gap.h#L588)). 
The advertising type should be `advertising_type_t::CONNECTABLE_UNDIRECTED` and make sure the legacy PDU is used.

Reply:
That's what I did right? Did you have a look again?

Reply:
It's `LEGACY`, not `INVALID`. By "legacy" it means older BLE standard (before 5) without extended advertising, not the API being legacy